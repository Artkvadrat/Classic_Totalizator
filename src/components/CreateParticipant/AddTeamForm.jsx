import React from 'react';
import { Button, Form, Input } from 'antd';
import PropTypes from 'prop-types';

import AddingParameters from './AddingParameters';
import AddingPlayersInTeam from './AddingPlayersInTeam';

const AddTeamForm = ({ newTeam, addNewParameters, addData, submitHandler }) => {
  const canAddNewParameter = !newTeam.parameters[newTeam.parameters.length - 1]
    .type;
  const canAddNewPlayer = !newTeam.players[newTeam.players.length - 1].name;

  const changeNameHandler = (e) => {
    addData(e.target.value, 'changeTeamName');
  };

  const changePhotoLinkHandler = (e) => {
    addData(e.target.value, 'changeTeamPhotoLink');
  };

  const addParameter = () => addNewParameters('parameterForTeam');

  const addPlayer = () => addNewParameters('playerInTeam');

  const submitForm = () => {
    submitHandler('teamSubmit');
  };

  return (
    <Form layout="vertical" onFinish={submitForm}>
      <Form.Item
        label="Name"
        name="name"
        rules={[
          {
            required: true,
            message: 'Please enter name'
          }
        ]}
      >
        <Input onChange={changeNameHandler} />
      </Form.Item>

      <Form.Item
        label="Link to photo"
        name="photoLink"
        rules={[
          {
            required: true,
            message: 'Please enter link to photo'
          },
          {
            pattern: /.png$/,
            message: 'Enter .png photo please'
          }
        ]}
      >
        <Input onChange={changePhotoLinkHandler} />
      </Form.Item>

      <div style={{ textAlign: 'left' }}>
        <p>Players</p>
        {newTeam.players.map((item, id) => (
          <AddingPlayersInTeam
            key={id}
            initialData={item}
            id={id}
            addData={addData}
            type="teamPlayers"
          />
        ))}
      </div>

      <Form.Item>
        <Button onClick={addPlayer} disabled={canAddNewPlayer}>
          Add new player
        </Button>
      </Form.Item>

      <div style={{ textAlign: 'left' }}>
        <p>Parameters</p>
        {newTeam.parameters.map((item, id) => (
          <AddingParameters
            key={id}
            initialData={item}
            id={id}
            addData={addData}
            type="teamParameters"
          />
        ))}
      </div>

      <Form.Item>
        <Button onClick={addParameter} disabled={canAddNewParameter}>
          Add new parameter
        </Button>
      </Form.Item>

      <Form.Item>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

AddTeamForm.propTypes = {
  newTeam: PropTypes.object,
  addNewParameters: PropTypes.func,
  addData: PropTypes.func,
  submitHandler: PropTypes.func
};

export default AddTeamForm;
