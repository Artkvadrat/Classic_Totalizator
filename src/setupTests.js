import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
