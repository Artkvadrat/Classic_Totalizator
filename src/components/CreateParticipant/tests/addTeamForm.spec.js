import React from 'react';
import { shallow } from 'enzyme';
import { Form, Input, Button } from 'antd';

import '../../../setupTests';

import AddTeamForm from '../AddTeamForm';

describe('AddTeamForm component', () => {
  let wrapper;

  beforeAll(() => {
    window.matchMedia =
      window.matchMedia ||
      function () {
        return {
          matches: false,
          addListener: () => {},
          removeListener: () => {}
        };
      };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render without error', () => {
    const props = {
      newTeam: {
        name: '',
        players: [{ name: '' }],
        photoLink: '',
        parameters: [{ type: '', value: '' }]
      },
      addNewParameters: jest.fn(),
      addData: jest.fn(),
      submitHandler: jest.fn()
    };
    wrapper = shallow(<AddTeamForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render 2 parameters and players forms without error', () => {
    const props = {
      newTeam: {
        name: '',
        players: [{ name: '' }, { name: '' }],
        photoLink: '',
        parameters: [
          { type: '', value: '' },
          { type: '', value: '' }
        ]
      },
      addNewParameters: jest.fn(),
      addData: jest.fn(),
      submitHandler: jest.fn()
    };
    wrapper = shallow(<AddTeamForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle changes in inputs', () => {
    const props = {
      newTeam: {
        name: '',
        players: [{ name: '' }],
        photoLink: '',
        parameters: [{ type: '', value: '' }]
      },
      addNewParameters: jest.fn(),
      addData: jest.fn(),
      submitHandler: jest.fn()
    };
    wrapper = shallow(<AddTeamForm {...props} />);

    const fakeEvent = {
      target: { value: 'test' }
    };

    const nameInput = wrapper.find(Input).at(0);
    nameInput.simulate('change', fakeEvent);
    expect(props.addData).toHaveBeenCalled();
    expect(props.addData).toHaveBeenCalledTimes(1);
    expect(props.addData).toHaveBeenCalledWith('test', 'changeTeamName');

    const photoLinkInput = wrapper.find(Input).at(1);
    photoLinkInput.simulate('change', fakeEvent);
    expect(props.addData).toHaveBeenCalled();
    expect(props.addData).toHaveBeenCalledTimes(2);
    expect(props.addData).toHaveBeenCalledWith('test', 'changeTeamPhotoLink');
  });

  it('should handle adding new player and parameter', () => {
    const props = {
      newTeam: {
        name: '',
        players: [{ name: '' }],
        photoLink: '',
        parameters: [{ type: '', value: '' }]
      },
      addNewParameters: jest.fn(),
      addData: jest.fn(),
      submitHandler: jest.fn()
    };
    wrapper = shallow(<AddTeamForm {...props} />);

    const firstButton = wrapper.find(Button).at(0);
    firstButton.simulate('click');
    expect(props.addNewParameters).toHaveBeenCalled();
    expect(props.addNewParameters).toHaveBeenCalledTimes(1);
    expect(props.addNewParameters).toHaveBeenCalledWith('playerInTeam');

    const secondButton = wrapper.find(Button).at(1);
    secondButton.simulate('click');
    expect(props.addNewParameters).toHaveBeenCalled();
    expect(props.addNewParameters).toHaveBeenCalledTimes(2);
    expect(props.addNewParameters).toHaveBeenCalledWith('parameterForTeam');
  });

  it('should handle submit form', () => {
    const props = {
      newTeam: {
        name: '',
        players: [{ name: '' }],
        photoLink: '',
        parameters: [{ type: '', value: '' }]
      },
      addNewParameters: jest.fn(),
      addData: jest.fn(),
      submitHandler: jest.fn()
    };
    wrapper = shallow(<AddTeamForm {...props} />);

    const form = wrapper.find(Form);
    form.simulate('finish');
    expect(props.submitHandler).toHaveBeenCalled();
    expect(props.submitHandler).toHaveBeenCalledTimes(1);
    expect(props.submitHandler).toHaveBeenCalledWith('teamSubmit');
  });
});
