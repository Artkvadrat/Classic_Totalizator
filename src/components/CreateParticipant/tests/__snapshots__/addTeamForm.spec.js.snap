// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTeamForm component should render 2 parameters and players forms without error 1`] = `
<ForwardRef(InternalForm)
  layout="vertical"
  onFinish={[Function]}
>
  <FormItem
    label="Name"
    name="name"
    rules={
      Array [
        Object {
          "message": "Please enter name",
          "required": true,
        },
      ]
    }
  >
    <Input
      onChange={[Function]}
      type="text"
    />
  </FormItem>
  <FormItem
    label="Link to photo"
    name="photoLink"
    rules={
      Array [
        Object {
          "message": "Please enter link to photo",
          "required": true,
        },
        Object {
          "message": "Enter .png photo please",
          "pattern": /\\.png\\$/,
        },
      ]
    }
  >
    <Input
      onChange={[Function]}
      type="text"
    />
  </FormItem>
  <div
    style={
      Object {
        "textAlign": "left",
      }
    }
  >
    <p>
      Players
    </p>
    <AddingPlayersInTeam
      addData={[MockFunction]}
      id={0}
      initialData={
        Object {
          "name": "",
        }
      }
      key="0"
      type="teamPlayers"
    />
    <AddingPlayersInTeam
      addData={[MockFunction]}
      id={1}
      initialData={
        Object {
          "name": "",
        }
      }
      key="1"
      type="teamPlayers"
    />
  </div>
  <FormItem>
    <Button
      disabled={true}
      onClick={[Function]}
    >
      Add new player
    </Button>
  </FormItem>
  <div
    style={
      Object {
        "textAlign": "left",
      }
    }
  >
    <p>
      Parameters
    </p>
    <AddingParameters
      addData={[MockFunction]}
      id={0}
      initialData={
        Object {
          "type": "",
          "value": "",
        }
      }
      key="0"
      type="teamParameters"
    />
    <AddingParameters
      addData={[MockFunction]}
      id={1}
      initialData={
        Object {
          "type": "",
          "value": "",
        }
      }
      key="1"
      type="teamParameters"
    />
  </div>
  <FormItem>
    <Button
      disabled={true}
      onClick={[Function]}
    >
      Add new parameter
    </Button>
  </FormItem>
  <FormItem>
    <Button
      htmlType="submit"
      type="primary"
    >
      Submit
    </Button>
  </FormItem>
</ForwardRef(InternalForm)>
`;

exports[`AddTeamForm component should render without error 1`] = `
<ForwardRef(InternalForm)
  layout="vertical"
  onFinish={[Function]}
>
  <FormItem
    label="Name"
    name="name"
    rules={
      Array [
        Object {
          "message": "Please enter name",
          "required": true,
        },
      ]
    }
  >
    <Input
      onChange={[Function]}
      type="text"
    />
  </FormItem>
  <FormItem
    label="Link to photo"
    name="photoLink"
    rules={
      Array [
        Object {
          "message": "Please enter link to photo",
          "required": true,
        },
        Object {
          "message": "Enter .png photo please",
          "pattern": /\\.png\\$/,
        },
      ]
    }
  >
    <Input
      onChange={[Function]}
      type="text"
    />
  </FormItem>
  <div
    style={
      Object {
        "textAlign": "left",
      }
    }
  >
    <p>
      Players
    </p>
    <AddingPlayersInTeam
      addData={[MockFunction]}
      id={0}
      initialData={
        Object {
          "name": "",
        }
      }
      key="0"
      type="teamPlayers"
    />
  </div>
  <FormItem>
    <Button
      disabled={true}
      onClick={[Function]}
    >
      Add new player
    </Button>
  </FormItem>
  <div
    style={
      Object {
        "textAlign": "left",
      }
    }
  >
    <p>
      Parameters
    </p>
    <AddingParameters
      addData={[MockFunction]}
      id={0}
      initialData={
        Object {
          "type": "",
          "value": "",
        }
      }
      key="0"
      type="teamParameters"
    />
  </div>
  <FormItem>
    <Button
      disabled={true}
      onClick={[Function]}
    >
      Add new parameter
    </Button>
  </FormItem>
  <FormItem>
    <Button
      htmlType="submit"
      type="primary"
    >
      Submit
    </Button>
  </FormItem>
</ForwardRef(InternalForm)>
`;
